### PR TITLE
[BEAM-6094] Add loopback environment for java.

### PR DESCRIPTION
**Pull request by robertwb**
------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

It will help us expedite review of your Pull Request if you tag someone (e.g. `@username`) to look at it.

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/) [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) </br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/) | --- | --- | ---

### CODE REVIEW

File: runners/reference/java/src/main/java/org/apache/beam/runners/reference/ExternalWorkerService.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+                    request.getLoggingEndpoint(),
+                    request.getControlEndpoint());
+                LOG.info("Successfully started worker {}.", request.getWorkerId());
+              } catch (Exception exn) {
+                LOG.error(String.format("Failed to start worker %s.", request.getWorkerId()), exn);
+              }
+            });
+    th.setName("SDK-worker-" + request.getWorkerId());
+    th.setDaemon(true);
+    th.start();
```
[the remark this answers is not in the document]
Comment (robertwb) [reply runners/reference/java/src/main/java/org/apache/beam/runners/reference/ExternalWorkerService.java]: It should be shut down via the protocol.

**Comment by robertwb** [comment]
Thanks! Yes, this will be tested in integration tests. (I thought
about it some and didn't see a good way to write (meaningful) unit
tests for it.)

**Comment by robertwb** [comment]
Run Python PreCommit

**Comment by robertwb** [comment]
Pre-existing (unrelated) failure in :beam-sdks-java-io-hadoop-format:test for Java, no Python changes.

**Comment by kennknowles** [comment]
@robertwb @mxm I just came across this resolving conflicts on the `build.gradle`. Why is the reference runner depending on a specific harness? Avoiding that dependency seems like probably basic hygiene that we'd want to maintain.

This PR has no description of what it is doing or why, so even after tracking down the source of the suspicious change, I can't really tell if it might actually be justified.

Just tagging the subject with a Jira is insufficient. It says nothing about what this implementation approach is or why it was chosen. Ideally, following the Jira would only be a "dig deeper" option while the PR description basically stands on its own.

And when I do copy/paste to look at the Jira it has no information whatsoever!

**Comment by mxm** [comment]
Fully agree, the JIRA issue, the PR description, and the commit should contain relevant information to understand the impact of the changes. I think this was skipped here due to implicit knowledge transfer while discussing/working on portability. Of course, this is no excuse.

This is for an External Environment which was merged here: https://github.com/apache/beam/pull/7078 An External Environment is used for an externally managed setup of the SDK harness, e.g. in a Kubernetes or Mesos context. The externally managed standby harness starts workers as they are requested by the Runner.

You are right that the Runner shouldn't take a dependency on a specific harness. I think we can remedy this by moving the code to `fn-execution` instead.

**Pull request merged**